content/05.main-text.md

```diff
@@ -257,8 +257,51 @@ We converted Hetionet v1.0 to HetMat format and uploaded the `hetionet-v1.0.hetm
 
 ### Permuted hetnets
 
+In order to generate a null distribution for a DWPC, we rely on DWPCs computed from permuted hetnets.
+We derive permuted hetnets from the unpermuted network using the XSwap algorithm [@10.1137/1.9781611972795.67].
+XSwap randomizes edges while preserving node degree.
+Therefore, it's ideal for generating null distributions that retain general degree effects,
+but destroy the actual meaning of edges.
+We adapt XSwap to hetnets by applying it separately to each metaedge [@rephetio; @doi:10.15363/thinklab.d178; @xswap].
+
+Project Rephetio created 5 permuted hetnets [@rephetio; @doi:10.15363/thinklab.d178],
+which were used to generate a null distribution of classifier performance for each metapath-based feature.
+Here, we aim to create a null distribution for individual DWPCs, which requires vastly more permuted values to estimate with accuracy.
+Therefore, we generated 200 permuted hetnets ([archive](https://github.com/greenelab/hetmech/blob/042063fb559048e52b3dc2731b6d6c6836f698cf/data/hetionet-v1.0-permutations.zip)).
+As part of this research, we developed the `xswap` Python [package](https://github.com/greenelab/xswap) with an optimized C implementation for network permutation [@xswap].
+
 ### Degree-grouping of node pairs
 
+For each of the 200 permuted networks and each of the 2,205 metapaths, we computed the entire DWPC matrix (i.e. all source nodes × target nodes).
+Therefore, for each actual DWPC value, we computed 200 permuted DWPC values.
+But 200 is too few to reliably model the null distribution, given the zero-inflated nature of DWPC distributions.
+Therefore, we grouped all permuted DWPCs with the same combination of source and target degrees.
```

Comment:
Just thinking, grouping isn't _just_ a matter of convenience or need---they are genuinely equivalent to additional permutations (ie. not just an approximation). Perhaps something more like this?

```suggestion
Because permutation preserves only node degree, DWPC values among nodes with the same source and target degrees are equivalent to additional permutations.
We greatly increased the effective number of permutations by grouping DWPC values according to node degree, affording us a superior estimation of the DWPC null distribution.
```

Reply:
> grouping isn't just a matter of convenience or need---they are genuinely equivalent to additional permutations (ie. not just an approximation)

essential point! Thanks for phrasing it so elegantly.